### PR TITLE
Cleanup amount.js

### PR DESCRIPTION
**Pull request by clark800**
Behavioral changes:

remove allow_nan option (option to convert NaN to 0)
all overflow checking is done in canonicalize
to_text always returns a string ('NaN' instead of NaN)
from_quality disallows XRP/XRP quality

### CODE REVIEW

**Comment by sublimator** [comment]
Most of this LGTM, but I'd like to see this taken further really, along with some new
documentation/tests.

It's weird that we throw errors for IOU amounts that are out of bounds, while
setting NaN internally for native amounts. I'd rather see errors thrown
consistently.

The `json wire limits` map to the `network limits` in rippled, which are limits
on values as serialised. I think we DO want to honour these limits for any
amount that is serialised, it's just that the constant copied from rippled was
the wrong one.

If you look at the uses of the source in the rippled code, you'll see the
comment is potentially misleading:

```
    // Maximum native value supported by the code
    static const std::uint64_t cMaxNative  = 9000000000000000000ull;
```

And that we actually want this one:

```
    // Max native value on network.
    static const std::uint64_t cMaxNativeN = 100000000000000000ull;
```

Consider this:

```
  inline
  bool
  isLegalNet (STAmount const& value)
  {
      return ! value.native() || (value.mantissa() <= STAmount::cMaxNativeN);
  }
```

and

```

src/ripple/app/transactors/CreateOffer.cpp:
  476          STAmount saTakerGets = mTxn.getFieldAmount (sfTakerGets);
  477  
  478:         if (!isLegalNet (saTakerPays) || !isLegalNet (saTakerGets))
  479              return temBAD_AMOUNT;
  480  

src/ripple/app/transactors/Payment.cpp:
   82              " saDstAmount=" << saDstAmount.getFullText ();
   83  
   84:         if (!isLegalNet (saDstAmount) || !isLegalNet (maxSourceAmount))
   85              return temBAD_AMOUNT;
   86  

src/ripple/app/transactors/SetTrust.cpp:
   78          std::uint32_t uQualityOut (bQualityOut ? mTxn.getFieldU32 (sfQualityOut) : 0);
   79  
   80:         if (!isLegalNet (saLimitAmount))
   81              return temBAD_AMOUNT;
   82  

src/ripple/app/transactors/Transactor.cpp:
  106      STAmount saPaid = mTxn.getTransactionFee ();
  107  
  108:     if (!isLegalNet (saPaid))
  109          return temBAD_AMOUNT;
  110  

src/ripple/protocol/STAmount.h:
  327  inline
  328  bool
  329: isLegalNet (STAmount const& value)
  330  {
  331      return ! value.native() || (value.mantissa() <= STAmount::cMaxNativeN);

```

In any case, we need to make some kind of distinction between intermediate
amounts and final amounts, which will serialised. Do we want to always enforce
those limits as invariants in the constructor? Or in the serialisation code (to
json or bytes).

(While it does seem to do the right thing and throw, we don't seem to have  any
tests for binary serialisation code when an amount _value is NaN or
BigNumber(NaN))

Because amounts aren't arbitrary precision and have a floating point, there's
problems with operations including `insignificant` operands.

What's ripple-lib do for `12345678901234500000000000000  + 1234` ?

Rounding up there doesn't make sense. And silently nooping doesn't either.

```
it ('has no sense of the significance', function() {
  var amt = Amount.from_json('12344567890123450000000/USD/1');
  var amt2 = Amount.from_json('4999999/USD/1'); 
  var amt3 = amt.add(amt2);
  assert.strictEqual(amt.to_text(), amt3.to_text());
});
```

In summary:

0. Use correct native network limit
1. Tests for binary serialisation of invalid amounts
2. Consider uses of intermediate vs final amounts
3. Consider where invariants should be enforced
    1. Is there value in invalid Amount objects?
4. Consider whether to throw when an operation becomes a noop due 
   to an insignificant operand
6. Document/rationalise/justify behaviours, especially regarding default rounding modes

I'm sort of sold on encouraging the use of arbitrary arithmetic for intermediate results, and enforcing limit invariants in the constructors, forcing callers to be explicit about rounding/truncation etc, but I'm not sure if there's a realistic migration path. I haven't really thought it through, but I'd happily read someone else's thoughts.

**Comment by sublimator** [comment]
Do you want me to raise issues as comments here? On comments on RLJS-207, or as new JIRA issues?

File: src/js/ripple/amount.js [hunk 1]
```diff
@@ -574,40 +566,27 @@ Amount.prototype.parse_quality = function(quality, counterCurrency, counterIssue
...
+    throw new Error('XRP/XRP quality is not allowed');
   }
 
-  // Correct offset if xrp_as_drops option is not set and base currency is XRP
-  if (!opts.xrp_as_drops &&
-      baseCurrency.is_valid() &&
+  var adjusted;
+  var inversion = opts.inverse ? -1 : 1;
+  // Correct for XRP to drops factors
+  if (this._is_native) {
```
Comment (sublimator) [inline src/js/ripple/amount.js]: Shouldn't xrp_as_drops be honoured regardless of whether it's the base or counter currency ?

File: src/js/ripple/amount.js [hunk 2]
```diff
@@ -176,13 +189,7 @@ Amount.prototype.ratio_human = function(denominator, opts) {
...
 
   var numerator = this;
-
-  if (typeof denominator === 'number' && parseInt(denominator, 10) === denominator) {
-    // Special handling of integer arguments
-    denominator = Amount.from_json(String(denominator) + '.0');
-  } else {
-    denominator = Amount.from_json(denominator);
-  }
+  denominator = Amount.from_json(noIntegers(denominator));
```
Comment (sublimator) [inline src/js/ripple/amount.js]: The original tacking on of `.0` seems ill considered. 

Perhaps an IOU style amount should be created for these cases where `noIntegers` is used, otherwise the number is parsed as a native amount (which seem to be truncated to 6 decimal places (iirc))

File: src/js/ripple/amount.js [hunk 3]
```diff
@@ -325,7 +327,9 @@ Amount.prototype.invert = function() {
 
 Amount.prototype.canonicalize = function(roundingMode) {
   if (this._is_native) {
-    this._value = this._value.round(6, BigNumber.ROUND_DOWN);
+    var value = this._value.round(6, BigNumber.ROUND_DOWN);
+    var isOverflow = value.abs().greaterThan(Amount.bi_xrp_max);
+    this._value = isOverflow ? new BigNumber(NaN) : value;
```
Comment (sublimator) [inline src/js/ripple/amount.js]: Why not blow up for native amounts in check limits, as with the iou amounts?

File: src/js/ripple/amount.js (hunk 2 above)
Comment (sublimator) [inline src/js/ripple/amount.js]: I mean, it seems like from_json isn't desired, so it's not really wanting to interpret it as a native amount. But then later, I see code like equals() which will just use from_json. Hrmmm. Not really consistent semantics.

I can't recall off hand what from_json does for Number values.

Comment (sublimator) [inline src/js/ripple/amount.js]: No cleanup of amount.js would be complete without at least considering this:

```

>>> Amount.from_json(1.0).to_text()
'1'

>>> Amount.from_json(1.1).to_text()
'1100000'

>>> Amount.from_json(1).to_text()
'1'

>>> Amount.from_json('1.0').to_text()
'1000000'

>>> Amount.from_json('1.1').to_text()
'1100000'

>>> Amount.from_json('1').to_text()
'1'

```

File: test/amount-test.js [hunk 4]
```diff
@@ -330,8 +330,8 @@ describe('Amount', function() {
   });
   describe('Amount parsing', function() {
     it('Parse invalid string', function() {
-      assert.strictEqual(Amount.from_json('x').to_text(), '0');
-      assert.strictEqual(typeof Amount.from_json('x').to_text(true), 'number');
+      assert.strictEqual(Amount.from_json('x').to_text(), 'NaN');
+      assert.strictEqual(typeof Amount.from_json('x').to_text(true), 'string');
       assert(isNaN(Amount.from_json('x').to_text(true)));
```
Comment (sublimator) [inline test/amount-test.js]: Could remove the now ignored argument for to_text

Comment (clark800) [inline test/amount-test.js]: Will do

File: src/js/ripple/amount.js (hunk 2 above)
Comment (clark800) [inline src/js/ripple/amount.js]: Yeah, this should be fixed. I think it deserves it's own ticket and pull request because it will probably require a lot of changes to the tests: https://ripplelabs.atlassian.net/browse/RLJS-215

We need to be really careful with this though, I'm not familiar with all the things this change could break.

Comment (clark800) [inline src/js/ripple/amount.js]: > Perhaps an IOU style amount should be created for these cases where noIntegers is used, otherwise the number is parsed as a native amount

I think the code is trying to allow the denominator parameter to be a native amount, but forcing it to be interpreted as XRP rather than drops, as it would be without the integer check. This can also be fixed by https://ripplelabs.atlassian.net/browse/RLJS-215

File: src/js/ripple/amount.js (hunk 3 above)
Comment (clark800) [inline src/js/ripple/amount.js]: I agree, these should be consistent. The question is whether it is better to throw or use NaN. This problem applies throughout the Amount class; some methods throw, others return NaN. Throwing is nice because the client code finds out about the issue right away.

The counter-argument might be as follows. Let's say ripple-trade adds two amounts using the Amount class to display the sum on the page and they forget to check for exceptions. If we use NaN, then presumably the box on the page would show "NaN", whereas if we throw, it could break the whole page. Arguably, breaking the whole page is ok because technically it is a bug that the client is not checking for errors and it should be fixed immediately.

File: src/js/ripple/amount.js (hunk 1 above)
Comment (clark800) [inline src/js/ripple/amount.js]: That would make sense, but I'm not really sure what is going on here. The original code only had a check for the base currency. The tests work either way.

**Comment by clark800** [comment]
There are a ton of issues with amount.js and I think we should take it one step at a time. So I think we should do new JIRA tasks for any "big" changes. I'd define "big" as something that would require more than a few minutes to change. I made a couple JIRA tasks:
https://ripplelabs.atlassian.net/browse/RLJS-215
https://ripplelabs.atlassian.net/browse/RLJS-217

**Comment by clark800** [comment]
> Use correct native network limit

Why is the network limit preferable to the code limit? For example if you go over the network limit with an intermediate value, but are back under before serializing, then there shouldn't be an error. So I would think we want to leave it as is.

> Tests for binary serialisation of invalid amounts
> Is there value in invalid Amount objects?

Not that I can think of. I think it might be best to always throw. https://ripplelabs.atlassian.net/browse/RLJS-217
The tests won't be necessary if amounts are always valid. 

> Consider uses of intermediate vs final amounts
> I'm sort of sold on encouraging the use of arbitrary arithmetic for intermediate results

I think the idea is to emulate the behavior of rippled's arithmetic, which is not arbitrary precision if I understand correctly. So if that's the goal, then we can't use arbitrary precision for intermediate results.

> Consider where invariants should be enforced

I think amount.js is doing this correctly already, but it looks like we might need to add a check to serializedtype.js. Should we throw in STAmount.serialize if the value is greater than cMaxNativeN = 100000000000000000ull?

> Consider whether to throw when an operation becomes a noop due to an insignificant operand

This would break compatibility with rippled, so probably not what we want.

> Document/rationalise/justify behaviours, especially regarding default rounding modes

Can you be more specific?

File: src/js/ripple/amount.js (hunk 3 above)
Comment (sublimator) [inline src/js/ripple/amount.js]: Yep, strong trees grow in windy areas.

**Comment by sublimator** [comment]
> The tests won't be necessary if amounts are always valid.

Maybe, but tests are useful for many reasons: documentation, regression proofing, declaring intent twice (like a password confirm), yada, yada

> This (throwing) would break compatibility with rippled,

I would argue rippled is broken in this respect, and if max compat weren't a concern, I'd most definitely throw.

It gets me to wondering why we don't replicate all the rounding quirks that were mentioned at one point if compatibility is a goal? How much compat, to what end?

> Should we throw in STAmount.serialize if the value is greater than cMaxNativeN = 100000000000000000ull?

I would say yes. No good could ever come from such an amount.

Also, see rippled's canonicalize:

```
src/ripple/protocol/impl/STAmount.cpp:
  784          }
  785  
  786:         if (mValue > cMaxNativeN)   <-----  network limit
  787              throw std::runtime_error ("Native currency amount out of range");
  788  
```

>>> Document/rationalise/justify behaviours, especially regarding default rounding modes

Pro-actively, systematically, and for the ages (comments in the code) describe/ justify how all this works.

So if I'm understanding you right, you believe we should honour rippled limits on both intermediate and final values? And you are in favour of removing the ability to construct invalid amounts, but have some concerns about temporary pains this may cause?

**Comment by clark800** [comment]
> tests are useful for many reasons

I mean that if we throw on every error, there won't be any invalid amounts to test. We would only be able to test that Amount throws on each possible error.

> why we don't replicate all the rounding quirks

Because the one rounding quirk that we diverge on is actually a bug in rippled which will presumably be fixed in the future. The rest of the quirks are not thought to be bugs in rippled.

> I would say yes.

Will do.

> justify how all this works.

I'd prefer to focus on fixing the issues first.

> you believe we should honour rippled limits on both intermediate and final values?

Yes, that is required for compatibility.

> And you are in favour of removing the ability to construct invalid amounts, but have some concerns about temporary pains this may cause?

Yes, but not just constructing them, also creating them through overflow etc.

**Comment by sublimator** [comment]
So in the long term, 100% compatibility is a goal?

>>> I'd prefer to focus on fixing the issues first.

Clarity = power.

I know I'd like to see it, as someone reviewing this, to help me make sense
of all the changes/fixes, and it shows me this has been well thought
through.

For example, why is the noop/insignificance thing not considered a bug?

**Comment by clark800** [comment]
> So in the long term, 100% compatibility is a goal?

It seems like some people are saying so. I'm not convinced that it is necessary. David Schwartz is not convinced either. If you want to discuss this, it is probably better to talk in the slack channel.

> I'd like to see it

It doesn't make sense to document when things are changing so fast. After fixing everything there shouldn't be too much left that needs documentation because it will all just make sense.

> why is the noop/insignificance thing not considered a bug?

That is just how fixed precision arithmetic works.

**Comment by sublimator** [comment]
> That is just how fixed precision arithmetic works.

Sure, but many decimal libraries will report an error when there's not enough precision to describe a certain amount, and you have to be explicit about how you round/truncate.

**Comment by sublimator** [comment]
> It doesn't make sense to document when things are changing so fast.

Yet I'm looking at a pull request where IOU limits exceeded are punished with errors and native amounts become `NaN`.  So you can understand why I'd like to see some better explanation.

**Comment by clark800** [comment]
> Yet I'm looking at a pull request where IOU limits exceeded are punished with errors and native amounts become NaN. So you can understand why I'd like to see some better explanation.

That was the pre-existing behavior. It is bad, it will be removed. Why would we document something just to delete that documentation shortly thereafter?

**Comment by clark800** [comment]
> Sure, but many decimal libraries will report an error when there's not enough precision to describe a certain amount, and you have to be explicit about how you round/truncate.

In any event, if you are going to break compatibility with rippled, you might as well go for the arbitrary precision option.

**Comment by sublimator** [comment]
You need to serialise amounts to talk to rippled at some point

**Comment by clark800** [comment]
I don't think throwing on precision errors is a good option. First of all, it is hard to define when to throw. Assume you have 3 digits of precision and you try to add 1.0 + 0.123, the answer is 1.12; should you throw because you lost the last digit? Or should you throw only if you lose all the digits in one term? If so, what about 100.0 + 0.5 = 101? This case is off from the real sum by 0.5, which is equally off as if the 0.5 was dropped, but in the opposite direction due to rounding. So despite being equally off, one case throws and the other doesn't? And what about precision loss during multiplication and division?

And the exception would only be useful if client code has something useful to do in response.

**Comment by sublimator** [comment]
>  it is hard to define when to throw. 
> And what about precision loss during multiplication and division?

Yeah, division especially (repeats) would need rounding modes.

I think in most cases, when adding/subtracting for like currency, most terms will be compatible. But because rippled uses fixed precision, but floating point, when they aren't compatible, I don't believe we should make guesses as to what should be done. Callers should have to respond to errors, or proactively choose what to do with spills prior. 

The 3 digit precision mental experiments aren't really on point, because we have 16, but in both addition cases there, I guess I'm suggesting throwing.

Maybe these spills really are insignificant, and can be safely ignored and rounded away. (After all, it must be done for 1 divide by 3). I'm not convinced yet though.

> The exception would only be useful if client code has something useful to do in response.

Not doing anything can sometimes be useful.

File: src/js/ripple/amount.js (hunk 2 above)
Comment (sublimator) [inline src/js/ripple/amount.js]: Is this noIntegers used primarily in the _human family? If so, your interpretation would seem correct.

File: src/js/ripple/amount.js [hunk 5]
```diff
@@ -9,6 +9,19 @@ var Seed      = require('./seed').Seed;
 var Currency  = require('./currency').Currency;
 var BigNumber = require('./bignumber');
 
+
+function isInteger(number) {
+  return parseInt(number) === number;
+}
+
+function noIntegers(value) {
```
Comment (sublimator) [inline src/js/ripple/amount.js]: It feels kind of pointless commenting on this, given it's a transitional commit, but if this commit has any value, so does a comment, so here goes:

`noIntegers` ?

The number `1.0` IS an integer, as `isInteger` would tell you. 
You and I happen to know the purpose for this function, but other people I'm sure would benefit from a better name, and some explanatory comments.  Off the top of my head: `forceFloatRepr` or `forceDecimalPoint`

File: src/js/ripple/amount.js [hunk 6]
```diff
@@ -574,40 +567,30 @@ Amount.prototype.parse_quality = function(quality, counterCurrency, counterIssue
...
-  // Correct offset if xrp_as_drops option is not set and base currency is XRP
-  if (!opts.xrp_as_drops &&
-      baseCurrency.is_valid() &&
-      baseCurrency.is_native()) {
-    if (opts.inverse) {
-      power -= 1;
+  var adjusted;
+  var inversion = opts.inverse ? -1 : 1;
+  // Correct for XRP to drops factors
+  if (!opts.xrp_as_drops) {
```
Comment (sublimator) [inline src/js/ripple/amount.js]:
```javascript
  /*
  The quality, as stored in the last 64 bits of a directory index, is stored as
  the quotient of TakerPays/TakerGets. 

  When `opts.inverse` is true we are looking at a quality used for determining a
  `bid` price and it must first be inverted, before our declared base/counter
  currencies are in line with the price.

  For example:

    quality as stored :  5 USD          /  3000000 drops
    inverted          :  3000000 drops  /          5 USD
  */
  var adjusted = opts.inverse ? inverse(value) : value;
  var nativeAdjusted = adjusted;

  if (!opts.xrp_as_drops) {
    // `In a currency exchange, the exchange rate is quoted as the units of the
    //  counter currency in terms of a single unit of a base currency`. A
    //  quality is how much taker must `pay` to get ONE `gets` unit thus:
    //    pays ~= counterCurrency 
    //    gets ~= baseCurrency.
    if (this._is_native) {
      // pay:$price              drops  get:1 X
      // pay:($price / 1,000,000)  XRP  get:1 X
      nativeAdjusted = adjusted.div(Amount.bi_xns_unit);
    } else if (baseCurrency.is_valid() && baseCurrency.is_native()) {
      // pay:$price X                   get:1 drop
      // pay:($price * 1,000,000) X     get:1 XRP
      nativeAdjusted = adjusted.times(Amount.bi_xns_unit); 
    }
  }
  
  this._set_value(nativeAdjusted);
```

I still prefer this form, as I find it easier to read. That's debatable of course, but it uses measurably less steps.

File: test/amount-test.js [hunk 7]
```diff
@@ -1182,11 +1187,11 @@ describe('Amount', function() {
 
   describe('amount limits', function() {
     it ('max JSON wire limite', function() {
-      assert.strictEqual(Amount.bi_xns_max.toString(), '9000000000000000000');
+      assert.strictEqual(Amount.bi_xns_max.toString(), '100000000000000000');
```
Comment (sublimator) [inline test/amount-test.js]: There are unfortunately, some entries in old ledgers, that exceed this limit :( 

Monkey patching as configuration works I guess.

File: test/serializedtypes-test.js [hunk 8]
```diff
@@ -570,8 +572,10 @@ describe('Serialized types', function() {
       assert.strictEqual(types.Amount.parse(so).to_json(), '270544960');
     });
     it('Parse 1161981756646125568 XRP', function () {
-      var so = new SerializedObject('5020304050607080');
-      assert.strictEqual(types.Amount.parse(so).to_json(), '1161981756646125696');
+      assert.throws(function() {
+        var so = new SerializedObject('1161981756646125568');
```
Comment (sublimator) [inline test/serializedtypes-test.js]: The original line was `var so = new SerializedObject('5020304050607080');`  which doesn't seem to map to the `Parse 1161981756646125568 XRP'`. 

May as well have it make sense?

File: src/js/ripple/amount.js [hunk 9]
```diff
@@ -342,8 +343,12 @@ Amount.prototype._check_limits = function() {
...
-  if (!this._is_native) {
-    var absval = this._value.absoluteValue();
+  var absval = this._value.absoluteValue();
+  if (this._is_native) {
+    if (absval.greaterThan(Amount.bi_xrp_max)) {
+      throw new Error('Exceeding max value of ' + Amount.bi_xrp_max.toString());
+    }
+  } else {
     if (absval.lessThan((new BigNumber(Amount.min_value)).absoluteValue())) {
       throw new Error('Exceeding min value of ' + Amount.min_value);
```
Comment (sublimator) [inline src/js/ripple/amount.js]: It might be worth making some `bi_min_value` and `bi_max_value` BigNumber constants, given we do these checks a lot ?

File: src/js/ripple/amount.js (hunk 2 above)
Comment (sublimator) [inline src/js/ripple/amount.js]: It does indeed seem a *_human, force XRP scale for native amounts hack as you said.

**Comment by sublimator** [comment]
No real problems popping out, so I guess LGTM. 

It would be good if some other people had a look too though.

File: src/js/ripple/amount.js (hunk 9 above)
Comment (clark800) [inline src/js/ripple/amount.js]: Yeah, I was thinking the same thing. I was going to wait because I also want to unexpose all the bi_* symbols so that dependencies aren't tied to our internal bignumber library. But I think it's worth making this part of the change now.

File: test/amount-test.js (hunk 7 above)
Comment (clark800) [inline test/amount-test.js]: How is that possible? I thought this was the total number of XRP drops created at the beginning (100 billion XRP * 1 million drops/XRP = 1e11 * 1e6 = 1e17).

File: src/js/ripple/amount.js (hunk 5 above)
Comment (clark800) [inline src/js/ripple/amount.js]: Ok, using `ensureDecimalPoint`

File: src/js/ripple/amount.js (hunk 6 above)
Comment (clark800) [inline src/js/ripple/amount.js]: Ok, patch applied

File: test/serializedtypes-test.js (hunk 8 above)
Comment (clark800) [inline test/serializedtypes-test.js]: It looks like that was supposed to be the hex representation, but it was wrong. I'll change it to the actual hex representation and add a comment.

**Comment by wltsmrz** [comment]
Is this issue addressed? https://github.com/ripple/ripple-lib/issues/175

**Comment by clark800** [comment]
No, should it be? I'm not clear on what the desired behavior is.

**Comment by wltsmrz** [comment]
It doesn't have to be, but I noticed that there are some Amount-related issues. This one I believe is addressed in the current pull request https://github.com/ripple/ripple-lib/issues/100

I think the desired behavior is to make `is_valid` more strict

```
> require('./').Amount.is_valid('1/USD/asdf')
true
> require('./').Amount.is_valid_full('1/USD/asdf')
false

```

**Comment by clark800** [comment]
#100 is https://ripplelabs.atlassian.net/browse/RLJS-215

Does that mean renaming is_valid_full to is_valid and removing the original is_valid?

File: test/amount-test.js (hunk 7 above)
Comment (sublimator) [inline test/amount-test.js]: There's thing like old offers, with takerPays of some huge amount. Recall that offers need only be partially funded.

Comment (sublimator) [inline test/amount-test.js]: @shekenahglory

This change might be of concern to you when processing old offers.

File: src/js/ripple/amount.js [hunk 10]
```diff
@@ -975,29 +937,25 @@ Amount.prototype.not_equals_why = function(d, ignore_issuer) {
     return 'Not an Amount';
   }
 
-  var result = false;
-
   if (!this.is_valid() || !d.is_valid()) {
-    result = 'Invalid amount.';
+    return 'Invalid amount.';
```
Comment (geertweening) [inline src/js/ripple/amount.js]: No need for `else` if you're using returns. Personally I like the old approach of using a `result` var that gets returned in the end, but early return is fine with me as well. If you're using returns though, don't use `else`'s. Makes it confusing

**Comment by geertweening** [comment]
LGTM

File: src/js/ripple/amount.js (hunk 10 above)
Comment (clark800) [inline src/js/ripple/amount.js]: I switched to returns because of the long-term goal of minimizing the use of variables (one of the todo items on the pull request for the javascript style guide). I'll refactor to remove else's.
